crates/cli-support/src/js/jsdoc.rs

```diff
@@ -0,0 +1,970 @@
+                return None;
+            };
+            text = trim_left_space(&text[name.len()..]);
+
+            let mut default = None;
+            if text.starts_with('=') {
+                text = trim_left_space(&text[1..]);
+                // the default value doesn't have to be a valid JS expression,
+                // so we just search for ']', '\n', or end of string
+                let end = text.find([']', '\n']).unwrap_or(text.len());
+                let default_text = text[..end].trim();
+                if !default_text.is_empty() {
+                    default = Some(default_text.to_string());
+                }
+
+                text = &text[end..];
+                if !text.is_empty() {
+                    text = trim_left_space(&text[1..]);
+                }
+            } else if text.starts_with(']') {
+                text = trim_left_space(&text[1..]);
+            }
+
+            (
+                default
+                    .map(Optionality::OptionalWithDefault)
+                    .unwrap_or(Optionality::Optional),
+                name.to_string(),
+            )
+        } else {
+            let Some(name) = consume_parameter_name_path(text) else {
+                // the name is not well-formed
+                return None;
+            };
+            text = trim_left_space(&text[name.len()..]);
+            (
+                if optional_by_type {
+                    Optionality::Optional
+                } else {
+                    Optionality::Required
+                },
+                name.to_string(),
+            )
+        };
+
+        Some(Self {
+            ty,
+            optional,
+            name,
+            description: text.to_string(),
+        })
+    }
+}
+
+impl ReturnsTag {
+    fn parse(rest: &str) -> Option<Self> {
+        // A bit careful now, because we want to keep the initial new lines of
+        // the description.
+        let mut text = {
+            let trimmed = trim_left(rest);
+            if trimmed.starts_with('{') {
+                trimmed
+            } else {
+                trim_left_space(rest)
+            }
+        };
+
+        let mut ty = None;
+        if text.starts_with('{') {
+            ty = consume_type_script_expression(&text[1..]).map(|t| t.to_string());
+
+            if let Some(ty) = &ty {
+                text = trim_left_space(&text[(ty.len() + 2)..]);
+            } else {
+                // the type expression is not terminated, so the tag is not well-formed
+                return None;
+            }
+        }
+        ty = post_process_typescript_expression(ty);
+
+        Some(Self {
+            ty,
+            description: text.to_string(),
+        })
+    }
+}
+
+fn post_process_typescript_expression(expr: Option<String>) -> Option<String> {
+    expr.and_then(|e| if e.trim().is_empty() { None } else { Some(e) })
+}
+
+/// A version trim_start that ignores text direction.
+fn trim_left(s: &str) -> &str {
```

Comment:
I made three functions for trimming, because I don't understand the text direction behavior of [`trim_start`](https://doc.rust-lang.org/std/primitive.str.html#method.trim_start). JSDoc is inherently left to right, so I don't know how this will interact with Arabic text.

Am I just paranoid here? Should I just use the std functions?

Reply:
The std documentation seems clear, even has some code examples. I don't think you are paranoid.